http/src/main/java/io/micronaut/http/uri/DefaultFormUrlEncodedDecoder.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.http.uri;
+
+import io.micronaut.context.annotation.Requires;
+import io.micronaut.http.form.FormConfiguration;
+import io.micronaut.http.form.FormUrlEncodedDecoder;
+import jakarta.inject.Singleton;
+import reactor.util.annotation.NonNull;
+import java.nio.charset.Charset;
+import java.util.Map;
+
+@Requires(missingBeans = FormUrlEncodedDecoder.class)
+@Singleton
+final class DefaultFormUrlEncodedDecoder implements FormUrlEncodedDecoder {
+    private final FormConfiguration formConfiguration;
+
+    DefaultFormUrlEncodedDecoder(FormConfiguration formConfiguration) {
+        this.formConfiguration = formConfiguration;
+    }
+
+    @Override
+    @NonNull
+    public Map<String, Object> decode(@NonNull String formUrlEncodedString, @NonNull Charset charset) {
+        QueryStringDecoder decoder = new QueryStringDecoder(formUrlEncodedString, charset, false, formConfiguration.getMaxDecodedKeyValueParameters());
```

Comment:
does a new instance need creating on each invocation of `decode` or can it be assigned to a final field of the type?

Reply:
Yes, I think you need to instantiate one instance and then invoke `QueryStringDecoder:parameters()` per invocation.